### PR TITLE
feat(linter): implement `useExplicitLengthCheck`

### PR DESCRIPTION
**Pull request by minht11**
## Summary
Closes https://github.com/biomejs/biome/issues/2564
Implement [unicorn/explicit-length-check](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/explicit-length-check.md).

Differences from unicorn:
1. unicorn checks `length`, `size`, but we additionally check `byteLength`, `byteOffset` properties.
2. This rule assumes that length property is always numeric, while unicorn sometimes can avoid false positives. Once biome has type inference this can be improved. In practice this should be rare.
3. `this.length` is checked in biome, but not in unicorn.
4. No options support.

## Test Plan
added tests and snapshots

### CODE REVIEW

File: crates/biome_js_syntax/src/expr_ext.rs [hunk 1]
```diff
@@ -1836,6 +1838,54 @@ impl TsStringLiteralType {
     }
 }
 
+/// Check if the SyntaxNode is a Negate Unary Expression
+/// ## Example
+/// ```js
+/// !x
+/// ```
+pub fn is_negation(node: &JsSyntaxNode) -> Option<JsUnaryExpression> {
```
Comment (minht11) [inline crates/biome_js_syntax/src/expr_ext.rs:1846]: I was not sure if this should be in `parser` or `syntaxt` crates.

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+            ),
+        );
+
+        let mut new_node = new_binary_expr.into_syntax();
+        let parent = state.node.syntax().parent()?;
+        // In cases like `export default!foo.length` -> `export default foo.length === 0`
+        // We need to add a space between keyword and expression
+        if matches!(
+            parent.kind(),
+            JsSyntaxKind::JS_EXPORT_DEFAULT_EXPRESSION_CLAUSE
```
Comment (minht11) [inline crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:263]: Is there a better way to check if space would be needed?

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 3]
```diff
@@ -0,0 +1,507 @@
...
+                | JsSyntaxKind::JS_AWAIT_EXPRESSION
+                | JsSyntaxKind::JS_IN_EXPRESSION
+                | JsSyntaxKind::JS_FOR_OF_STATEMENT
+                | JsSyntaxKind::JS_FOR_IN_STATEMENT
+                | JsSyntaxKind::JS_DO_WHILE_STATEMENT
+                | JsSyntaxKind::JS_CASE_CLAUSE
+        ) || does_unary_expr_needs_space(&parent)
+        {
+            // Make fake token to get leading trivia
+            let leading_trivia = make::token_decorated_with_space(T![=])
```
Comment (minht11) [inline crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:278]: I was not able to find a way to create trivia piecies standalone.

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+enum MemberPosition {
+    Left,
+    Right,
+}
+
+#[derive(PartialEq, Clone, Copy)]
+enum LengthCheck {
+    Zero,
+    NonZero,
+}
```
[the remark this answers is not in the document]
Comment (minht11) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs]: I replaced that whole expression into single match statement. Please take a look if it is better.

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+    Some((AnyJsExpression::cast(boolean_node?)?, is_negative))
+}
+
+/// Check if the SyntaxNode is a `Boolean` Call Expression
+/// ## Example
+/// ```js
+/// Boolean(x)
+/// ```
+pub fn is_boolean_call(node: &JsSyntaxNode) -> Option<JsCallExpression> {
```
[the remark this answers is not in the document]
Comment (minht11) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:439]: Not quite sure what you mean. is_boolean_constructor_call is only used in no_extra_boolean_cast.rs file. And that file already has is_boolean fn in which is different from the one that I have expr_ext. Tried replacing local one, but it tests failed.

File: crates/biome_js_syntax/src/expr_ext.rs (hunk 1 above)
Comment (Conaclos) [reply crates/biome_js_syntax/src/expr_ext.rs:1846]: `biome_js_syntax` is the good place :)

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Conaclos) [reply crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:439]: Thanks. Thus, don't take my comment into account.

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+    let (member_position, literal) = extract_binary_position_and_literal(&binary_expr)?;
+    let number = literal.as_js_number_literal_expression()?.as_number()?.round() as i64;
+
+    let length_check = match (member_position, binary_expr.operator().ok()?, number) {
+        // Zero length checks
+        // -------------------------
+        // `foo.length == 0`
+        (MemberPosition::Right, JsBinaryOperator::Equality, 0) |
+        // `0 == foo.length`
+        (MemberPosition::Left, JsBinaryOperator::Equality, 0) |
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:362]:
```suggestion
        // `0 == foo.length` or `foo.length == 0`
        (MemberPosition::Left | MemberPosition::Right, JsBinaryOperator::Equality, 0) |
```

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+        (MemberPosition::Left, JsBinaryOperator::GreaterThan, 1) |
+        // 0 === foo.length. Valid but we prefer right side to be a number
+        (MemberPosition::Left, JsBinaryOperator::StrictEquality, 0) => Some(LengthCheck::Zero),
+        // -------------------------
+        // Non-zero length checks
+        // -------------------------
+        // `foo.length !== 0`
+        (MemberPosition::Right, JsBinaryOperator::StrictInequality, 0) |
+        // `0 !== foo.length`
+        (MemberPosition::Left, JsBinaryOperator::StrictInequality, 0) |
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:375]:
```suggestion
        // `0 !== foo.length` or `foo.length !== 0`
        (MemberPosition::Left | MemberPosition::Right, JsBinaryOperator::StrictInequality, 0) |
```

File: crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+        // `foo.length != 0`
+        (MemberPosition::Right, JsBinaryOperator::Inequality, 0) |
+        // `0 != foo.length`
+        (MemberPosition::Left, JsBinaryOperator::Inequality, 0) |
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/lint/nursery/use_explicit_length_check.rs:379]:
```suggestion
        // `0 != foo.length` or `foo.length != 0`
        (MemberPosition::Left | MemberPosition::Right, JsBinaryOperator::Inequality, 0) |
```

**Review by Conaclos — APPROVED** [review]
I left some minor suggestions. Could you [add a changelog entry](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#changelog)? Thanks :)

**Comment by ematipico** [comment]
Can you resolve the conflicts @minht11? After that we can merge it

**Pull request merged**